src/GafferScene/SetExpression.cpp

```diff
@@ -0,0 +1,276 @@
+
+using namespace Gaffer;
+using namespace GafferScene;
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+#include <boost/config/warning_disable.hpp>
+#include <boost/spirit/include/qi.hpp>
+#include <boost/variant/recursive_variant.hpp>
+#include <boost/variant/apply_visitor.hpp>
+#include <boost/spirit/include/phoenix_operator.hpp>
+#include <boost/spirit/include/phoenix_function.hpp>
+
+#include <iostream>
+#include <vector>
+#include <string>
+
+namespace qi = boost::spirit::qi;
+namespace ascii = boost::spirit::ascii;
+
+// AST
+struct binary_op;
+struct nil {};
+
+struct ExpressionAst
+{
+	typedef
+	boost::variant<
+		nil, // can't happen!
+		std::string,
+		boost::recursive_wrapper<ExpressionAst>,
+		boost::recursive_wrapper<binary_op>
+		>
+	type;
+
+	ExpressionAst()
+		: expr(nil()) {}
+
+	template <typename Expr>
+	ExpressionAst(Expr const& expr)
+		: expr(expr) {}
+
+	ExpressionAst& operator+=(ExpressionAst const& rhs);
+	ExpressionAst& operator-=(ExpressionAst const& rhs);
+
+	type expr;
+};
+struct binary_op
+{
+	binary_op(
+		char op,
+		ExpressionAst const& left,
+		ExpressionAst const& right)
+		: op(op), left(left), right(right) {}
+
+	char op;
+	ExpressionAst left;
+	ExpressionAst right;
+};
+
+ExpressionAst& ExpressionAst::operator+=(ExpressionAst const& rhs)
+{
+	expr = binary_op('|', expr, rhs);
+	return *this;
+}
+
+ExpressionAst& ExpressionAst::operator-=(ExpressionAst const& rhs)
+{
+	expr = binary_op('&', expr, rhs);
+	return *this;
+}
+
+// Visiting the AST
+// ----------------
+struct AstPrinter
+{
+typedef void result_type;
+
+void operator()(qi::info::nil) const {}
+void operator()(std::string n) const { std::cout << n; }
+
+void operator()(ExpressionAst const& ast) const
+{
+	boost::apply_visitor(*this, ast.expr);
+}
+
+void operator()(binary_op const& expr) const
+{
+	std::cout << "op:" << expr.op << "(";
+	boost::apply_visitor(*this, expr.left.expr);
+	std::cout << ", ";
+	boost::apply_visitor(*this, expr.right.expr);
+	std::cout << ')';
+}
+
+};
+
+struct AstEvaluator
+{
+typedef PathMatcher result_type;
+
+AstEvaluator( const ScenePlug *scene )
+{
+	m_scene = scene;
+}
+
+result_type operator()(std::string setName) const
+{
+	// TODO: handle case in which the setName given is not a valid set
+	return m_scene->set( setName )->readable();
+}
+
+result_type operator()(ExpressionAst const& ast) const
+{
+	return boost::apply_visitor(*this, ast.expr);
+}
+
+result_type operator()(binary_op const& expr) const
+{
+	PathMatcher left = boost::apply_visitor(*this, expr.left.expr);
+	PathMatcher right = boost::apply_visitor(*this, expr.right.expr);
+
+	if( expr.op == '|' )
+	{
+		PathMatcher result = PathMatcher( left );
+		result.addPaths( right );
+		return result;
+	}
+	else if( expr.op == '&' )
+	{
+		PathMatcher result = PathMatcher();
+
+		for(PathMatcher::Iterator it = left.begin(); it != left.end(); ++it)
+		{
+			if( right.find( *it ) != right.end() )
```

Comment:
Maybe I didn't describe well enough (or maybe I'm wrong - let's find out!). You're correct that you only want to visit the actual match locations in the first set, the one you're iterating over. The problem I think we have is with the way we're looking up into the second set with `find()` - that is going to return a non-end iterator even if the location exists in the second set only as an ancestor. If I'm right, then including "/a/b/c" in `m1` in your unit test will give us a failure, because although "/a/b/c" doesn't exist in `m2` in its own right, it does exist there as an ancestor.

Either way, the optimisation I suggested (where we call match then `it.prune()` if there are no descendant matches) is well worth having...